Content.Server/Connection/ConnectionManager.Whitelist.cs

```diff
@@ -0,0 +1,241 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using Content.Server.Connection.Whitelist;
+using Content.Server.Connection.Whitelist.Conditions;
+using Content.Server.Database;
+using Content.Shared.CCVar;
+using Content.Shared.Database;
+using Content.Shared.Players.PlayTimeTracking;
+using Robust.Shared.Network;
+
+namespace Content.Server.Connection;
+
+/// <summary>
+/// Handles whitelist conditions for incoming connections.
+/// </summary>
+public sealed partial class ConnectionManager
+{
+    private PlayerConnectionWhitelistPrototype[]? _whitelists;
+
+    public void PostInit()
+    {
+        _cfg.OnValueChanged(CCVars.WhitelistPrototypeList, UpdateWhitelists, true);
+    }
+
+    private void UpdateWhitelists(string s)
+    {
+        var list = new List<PlayerConnectionWhitelistPrototype>();
+        foreach (var id in s.Split(','))
+        {
+            if (_prototypeManager.TryIndex(id, out PlayerConnectionWhitelistPrototype? prototype))
+            {
+                list.Add(prototype);
+            }
+            else
+            {
+                _sawmill.Fatal($"Whitelist prototype {id} does not exist. Denying all connections.");
+                _whitelists = null; // Invalidate the list, causes deny on all connections.
+                return;
+            }
+        }
+
+        _whitelists = list.ToArray();
+    }
+
+    private bool IsValid(PlayerConnectionWhitelistPrototype whitelist, int playerCount)
+    {
+        return playerCount >= whitelist.MinimumPlayers && playerCount <= whitelist.MaximumPlayers;
+    }
+
+    public async Task<(bool isWhitelisted, string? denyMessage)> IsWhitelisted(PlayerConnectionWhitelistPrototype whitelist, NetUserData data, ISawmill sawmill)
+    {
+        Dictionary<NetUserId, List<IAdminRemarksRecord>> cacheRemarks = new();
+        Dictionary<NetUserId, List<PlayTime>> cachePlaytime = new();
```

Comment:
Why are you storing this keyed on user ID if this function only runs for a single user at once?